README.md
xmpp-rs
=======

What's this?
------------

A very much WIP rust XMPP library with the goals of being type-safe, well-tested and simple.

Contact
-------

There is an XMPP MUC for the discussion of this library, feel free to join! :)

[chat@xmpp.rs](xmpp:chat@xmpp.rs?join)

Can I use it yet?
-----------------

No, there's still a lot of work that needs to be done before this could be used for anything.

I mean, if you really wanted to, sure, but don't expect it to work yet.

What license is it under?
-------------------------

LGPLv3 or later. See `COPYING` and `COPYING.LESSER`.

License yadda yadda.
--------------------

  Copyright 2017, xmpp-rs contributors.

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
Version NEXT:
DATE  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - Client Certificate Management for SASL EXTERNAL (XEP-0257)
        - JID Prep (XEP-0328)
        - Client State Indication (XEP-0352)
        - OpenPGP for XMPP (XEP-0373)
        - Bookmarks 2 (This Time it's Serious) (XEP-0402)
        - Anonymous unique occupant identifiers for MUCs (XEP-0421)
    * Breaking changes:
        - Presence constructors now take Into<Jid> and assume Some.
    * Improvements:
        - CI: refactor, add caching

Version 0.15.0:
2019-09-06  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - XHTML-IM (XEP-0071)
        - User Tune (XEP-0118)
        - Bits of Binary (XEP-0231)
        - Message Carbons (XEP-0280)
    * Breaking changes:
        - Stop reexporting TryFrom and TryInto, they are available in
          std::convert nowadays.
        - Bind has been split into BindQuery and BindResponse.
    * Improvements:
        - New DOAP file for a machine-readable description of the features.
        - Add various parser and formatter helpers on Hash.

Version 0.14.0:
2019-07-13  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>, Maxime “pep” Buquet <pep@bouah.net>
    * New parsers/serialisers:
        - Entity Time (XEP-0202).
    * Improvements:
        - Microblog NS (XEP-0227).
        - Update jid-rs dependency with jid split change (Jid, FullJid,
          BareJid) and reexport them.
        - Fix rustdoc options in Cargo.toml for docs.rs
    * Breaking changes:
        - Presence's show attribute is now Option<Show> and Show::None is no
          more.

Version 0.13.1:
2019-04-12  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Bugfixes:
        - Fix invalid serialisation of priority in presence.
        - Bump image size to u16 from u8, as per XEP-0084 version 1.1.2.
    * Improvements:
        - Drop try_from dependency, as std::convert::TryFrom got
          stabilised.

Version 0.13.0:
2019-03-20  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - User Avatar (XEP-0084).
        - Contact Addresses for XMPP Services (XEP-0157).
        - Jingle RTP Sessions (XEP-0167).
        - Jingle ICE-UDP Transport Method (XEP-0176).
        - Use of DTLS-SRTP in Jingle Sessions (XEP-0320).
    * Breaking changes:
        - Make 'id' required on iq, as per RFC6120 §8.1.3.
        - Refactor PubSub to have more type-safety.
        - Treat FORM_TYPE as a special case in data forms, to avoid
          duplicating it into a field.
        - Add forgotten i18n to Jingle text element.
    * Improvements:
        - Add various helpers for hash representations.
        - Add helpers constructors for multiple extensions (disco, caps,
          pubsub, stanza_error).
        - Use Into<String> in more constructors.
        - Internal change on attribute declaration in macros.
        - Reexport missing try_from::TryInto.

Version 0.12.2:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Improvements:
        - Reexport missing util::error::Error and try_from::TryFrom.

Version 0.12.1:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Improvements:
        - Reexport missing JidParseError from the jid crate.

Version 0.12.0:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Breaking changes:
        - Update dependencies.
        - Switch to git, upstream is now available at
          https://gitlab.com/xmpp-rs/xmpp-parsers
        - Switch to Edition 2018, this removes support for rustc
          versions older than 1.31.
        - Implement support for XEP-0030 2.5rc3, relaxing the ordering
          of children in disco#info.
    * Improvements:
        - Test for struct size, to keep them known and avoid bloat.
        - Add various constructors to make the API easier to use.
        - Reexport Jid from the jid crate, to avoid any weird issue on
          using different incompatible versions of the same crate.
        - Add forgotten 'ask' attribute on roster item (thanks O01eg!).
        - Use cargo-fmt on the codebase, to lower the barrier of entry.
        - Add a disable-validation feature, disabling many checks
          xmpp-parsers is doing.  This should be used for software
          which want to let invalid XMPP pass through instead of being
          rejected as invalid (thanks Astro-!).

Version 0.11.1:
2018-09-20  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Improvements:
        - Document all of the modules.

Version 0.11.0:
2018-08-03  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Breaking changes:
        - Split Software Version (XEP-0092) into a query and response
          elements.
        - Split RSM (XEP-0059) into a query and response elements.
        - Fix type safety and spec issues in RSM and MAM (XEP-0313).
        - Remove item@node and EmptyItems from PubSub events
          (XEP-0060).
    * Improvements:
        - Document many additional modules.
        - Add the <failure/> SASL nonza, as well as the SCRAM-SHA-256
          and the two -PLUS mechanisms.

Version 0.10.0:
2018-07-31  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - Added <stream:stream>, SASL and bind (RFC6120) parsers.
        - Added a WebSocket <open/> (RFC7395) implementation.
        - Added a Jabber Component <handshake/> (XEP-0114).
        - Added support for User Nickname (XEP-0172).
        - Added support for Stream Management (XEP-0198).
        - Added support for Bookmarks (XEP-0048).
        - Publish-Subscribe (XEP-0060) now supports requests in
          addition to events.
    * Breaking changes:
        - Switch from std::error to failure to report better errors.
        - Bump to minidom 0.9.1, and reexport minidom::Element.
    * Improvements:
        - Add getters for the best body and subject in message, to make
          it easier to determine which one the user wants based on

src/ns.rs
// Copyright (c) 2017-2018 Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
// Copyright (c) 2017 Maxime “pep” Buquet <pep@bouah.net>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const JABBER_CLIENT: &str = "jabber:client";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const XMPP_STANZAS: &str = "urn:ietf:params:xml:ns:xmpp-stanzas";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const STREAM: &str = "http://etherx.jabber.org/streams";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const SASL: &str = "urn:ietf:params:xml:ns:xmpp-sasl";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const BIND: &str = "urn:ietf:params:xml:ns:xmpp-bind";

/// RFC 6121: Extensible Messaging and Presence Protocol (XMPP): Instant Messaging and Presence
pub const ROSTER: &str = "jabber:iq:roster";

/// RFC 7395: An Extensible Messaging and Presence Protocol (XMPP) Subprotocol for WebSocket
pub const WEBSOCKET: &str = "urn:ietf:params:xml:ns:xmpp-framing";

/// XEP-0004: Data Forms
pub const DATA_FORMS: &str = "jabber:x:data";

/// XEP-0030: Service Discovery
pub const DISCO_INFO: &str = "http://jabber.org/protocol/disco#info";
/// XEP-0030: Service Discovery
pub const DISCO_ITEMS: &str = "http://jabber.org/protocol/disco#items";

/// XEP-0045: Multi-User Chat
pub const MUC: &str = "http://jabber.org/protocol/muc";
/// XEP-0045: Multi-User Chat
pub const MUC_USER: &str = "http://jabber.org/protocol/muc#user";

/// XEP-0047: In-Band Bytestreams
pub const IBB: &str = "http://jabber.org/protocol/ibb";

/// XEP-0048: Bookmarks
pub const BOOKMARKS: &str = "storage:bookmarks";

/// XEP-0059: Result Set Management
pub const RSM: &str = "http://jabber.org/protocol/rsm";

/// XEP-0060: Publish-Subscribe
pub const PUBSUB: &str = "http://jabber.org/protocol/pubsub";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_ERRORS: &str = "http://jabber.org/protocol/pubsub#errors";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_EVENT: &str = "http://jabber.org/protocol/pubsub#event";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_OWNER: &str = "http://jabber.org/protocol/pubsub#owner";

/// XEP-0071: XHTML-IM
pub const XHTML_IM: &str = "http://jabber.org/protocol/xhtml-im";
/// XEP-0071: XHTML-IM
pub const XHTML: &str = "http://www.w3.org/1999/xhtml";

/// XEP-0077: In-Band Registration
pub const REGISTER: &str = "jabber:iq:register";

/// XEP-0084: User Avatar
pub const AVATAR_DATA: &str = "urn:xmpp:avatar:data";
/// XEP-0084: User Avatar
pub const AVATAR_METADATA: &str = "urn:xmpp:avatar:metadata";

/// XEP-0085: Chat State Notifications
pub const CHATSTATES: &str = "http://jabber.org/protocol/chatstates";

/// XEP-0092: Software Version
pub const VERSION: &str = "jabber:iq:version";

/// XEP-0107: User Mood
pub const MOOD: &str = "http://jabber.org/protocol/mood";

/// XEP-0114: Jabber Component Protocol
pub const COMPONENT_ACCEPT: &str = "jabber:component:accept";

/// XEP-0114: Jabber Component Protocol
pub const COMPONENT: &str = "jabber:component:accept";

/// XEP-0115: Entity Capabilities
pub const CAPS: &str = "http://jabber.org/protocol/caps";

/// XEP-0118: User Tune
pub const TUNE: &str = "http://jabber.org/protocol/tune";

/// XEP-0157: Contact Addresses for XMPP Services
pub const SERVER_INFO: &str = "http://jabber.org/network/serverinfo";

/// XEP-0166: Jingle
pub const JINGLE: &str = "urn:xmpp:jingle:1";

/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP: &str = "urn:xmpp:jingle:apps:rtp:1";
/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP_AUDIO: &str = "urn:xmpp:jingle:apps:rtp:audio";
/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP_VIDEO: &str = "urn:xmpp:jingle:apps:rtp:video";

/// XEP-0172: User Nickname
pub const NICK: &str = "http://jabber.org/protocol/nick";

/// XEP-0176: Jingle ICE-UDP Transport Method
pub const JINGLE_ICE_UDP: &str = "urn:xmpp:jingle:transports:ice-udp:1";

/// XEP-0184: Message Delivery Receipts
pub const RECEIPTS: &str = "urn:xmpp:receipts";

/// XEP-0191: Blocking Command
pub const BLOCKING: &str = "urn:xmpp:blocking";
/// XEP-0191: Blocking Command
pub const BLOCKING_ERRORS: &str = "urn:xmpp:blocking:errors";

/// XEP-0198: Stream Management
pub const SM: &str = "urn:xmpp:sm:3";

/// XEP-0199: XMPP Ping
pub const PING: &str = "urn:xmpp:ping";

/// XEP-0202: Entity Time
pub const TIME: &str = "urn:xmpp:time";

/// XEP-0203: Delayed Delivery
pub const DELAY: &str = "urn:xmpp:delay";

/// XEP-0221: Data Forms Media Element
pub const MEDIA_ELEMENT: &str = "urn:xmpp:media-element";

/// XEP-0224: Attention
pub const ATTENTION: &str = "urn:xmpp:attention:0";

/// XEP-0231: Bits of Binary
pub const BOB: &str = "urn:xmpp:bob";

/// XEP-0234: Jingle File Transfer
pub const JINGLE_FT: &str = "urn:xmpp:jingle:apps:file-transfer:5";
/// XEP-0234: Jingle File Transfer
pub const JINGLE_FT_ERROR: &str = "urn:xmpp:jingle:apps:file-transfer:errors:0";

/// XEP-0257: Client Certificate Management for SASL EXTERNAL
pub const SASL_CERT: &str = "urn:xmpp:saslcert:1";

/// XEP-0260: Jingle SOCKS5 Bytestreams Transport Method
pub const JINGLE_S5B: &str = "urn:xmpp:jingle:transports:s5b:1";

/// XEP-0261: Jingle In-Band Bytestreams Transport Method
pub const JINGLE_IBB: &str = "urn:xmpp:jingle:transports:ibb:1";

/// XEP-0277: Microblogging over XMPP
pub const MICROBLOG: &str = "urn:xmpp:microblog:0";

/// XEP-0280: Message Carbons
pub const CARBONS: &str = "urn:xmpp:carbons:2";

/// XEP-0297: Stanza Forwarding
pub const FORWARD: &str = "urn:xmpp:forward:0";

/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASHES: &str = "urn:xmpp:hashes:2";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA_256: &str = "urn:xmpp:hash-function-text-names:sha-256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA_512: &str = "urn:xmpp:hash-function-text-names:sha-512";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA3_256: &str = "urn:xmpp:hash-function-text-names:sha3-256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA3_512: &str = "urn:xmpp:hash-function-text-names:sha3-512";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_BLAKE2B_256: &str = "urn:xmpp:hash-function-text-names:id-blake2b256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_BLAKE2B_512: &str = "urn:xmpp:hash-function-text-names:id-blake2b512";

/// XEP-0308: Last Message Correction
pub const MESSAGE_CORRECT: &str = "urn:xmpp:message-correct:0";

/// XEP-0313: Message Archive Management
pub const MAM: &str = "urn:xmpp:mam:2";

/// XEP-0319: Last User Interaction in Presence
pub const IDLE: &str = "urn:xmpp:idle:1";

/// XEP-0320: Use of DTLS-SRTP in Jingle Sessions
pub const JINGLE_DTLS: &str = "urn:xmpp:jingle:apps:dtls:0";

/// XEP-0328: JID Prep
pub const JID_PREP: &str = "urn:xmpp:jidprep:0";

/// XEP-0352: Client State Indication
pub const CSI: &str = "urn:xmpp:csi:0";

/// XEP-0353: Jingle Message Initiation
pub const JINGLE_MESSAGE: &str = "urn:xmpp:jingle-message:0";

/// XEP-0359: Unique and Stable Stanza IDs
pub const SID: &str = "urn:xmpp:sid:0";

/// XEP-0373: OpenPGP for XMPP
pub const OX: &str = "urn:xmpp:openpgp:0";
/// XEP-0373: OpenPGP for XMPP
pub const OX_PUBKEYS: &str = "urn:xmpp:openpgp:0:public-keys";

/// XEP-0380: Explicit Message Encryption
pub const EME: &str = "urn:xmpp:eme:0";

/// XEP-0390: Entity Capabilities 2.0
pub const ECAPS2: &str = "urn:xmpp:caps";
/// XEP-0390: Entity Capabilities 2.0
pub const ECAPS2_OPTIMIZE: &str = "urn:xmpp:caps:optimize";

/// XEP-0402: Bookmarks 2 (This Time it's Serious)
pub const BOOKMARKS2: &str = "urn:xmpp:bookmarks:0";
/// XEP-0402: Bookmarks 2 (This Time it's Serious)
pub const BOOKMARKS2_COMPAT: &str = "urn:xmpp:bookmarks:0#compat";

/// XEP-0421: Anonymous unique occupant identifiers for MUCs
pub const OID: &str = "urn:xmpp:occupant-id:0";

/// Alias for the main namespace of the stream, that is "jabber:client" when
/// the component feature isn’t enabled.
#[cfg(not(feature = "component"))]
pub const DEFAULT_NS: &str = JABBER_CLIENT;

/// Alias for the main namespace of the stream, that is
/// "jabber:component:accept" when the component feature is enabled.
#[cfg(feature = "component")]
pub const DEFAULT_NS: &str = COMPONENT_ACCEPT;
